File: test-suite/golden-tests/closure-conformance/hosts.py
import os



def home_tick(path):
    with open(path, "ab") as f:
        f.write(b"x")
    return os.path.getsize(path)


def home_ident(x):
    return x


def home_mark(s):
    return s


def home_str_len(s):
    return len(s)


def home_add(x, y):
    return x + y


def home_fail(path):
    raise RuntimeError("failHome ran: " + path)


def home_mk(k):
    return lambda x: x + k


def home_big(n):
    return list(range(n))


def home_use_pure(f, x):
    return f(x)


def home_take_thunk(t):
    t()
    t()
    return t()



def away_ident(x):
    return x


def away_bump(path, x):
    with open(path, "ab") as f:
        f.write(b"x")
    return os.path.getsize(path)


def away_take_thunk(t):
    t()
    t()
    return t()


def away_drop_thunk(t):
    return 0


def away_use_record(h):
    h["run"]()
    h["run"]()
    return h["run"]() + h["tag"]


def away_use_list(xs):
    xs[0]()
    xs[0]()
    return xs[0]()


def away_use_tuple(t):
    t[0]()
    t[0]()
    return t[0]() + t[1]


def away_use_opt(t):
    if t is None:
        return 0
    t()
    t()
    return t()


def away_use_nested(hs):
    hs[0]["run"]()
    hs[0]["run"]()
    return hs[0]["run"]()


def away_use_callback(f, x):
    f(x)
    f(x)
    return f(x)


def away_use_pure(f, x):
    return f(x)


def away_use_arity2(f, x, y):
    return f(x, y)


def away_use_pinned(c, x):
    c["inc"](x)
    c["inc"](x)
    return c["inc"](x)


def away_use_hof(f, g):
    return f(g)


def away_use_list_pure(fs, x):
    return sum(f(x) for f in fs)


def away_use_mk(f, a, b):
    return f(a)(b)


def away_use_thunk2(t):
    inner = t()
    inner()
    inner()
    return inner()


def away_use_thunk_fn(t, x):
    return t()(x)


def away_use_fn_thunk(f, t):
    return f(t)


def away_pass_through(f):
    return f


def away_map_away(f, xs):
    return [f(x) for x in xs]


def away_use_hof_eff(f, g):
    return f(g)


def away_use_hof2(f):
    return f(lambda g: g(3))


def away_two(a, b, x):
    with open(a, "ab") as fh:
        fh.write(b"x")
    return os.path.getsize(a) + len(b)


def away_mix(x, a):
    with open(a, "ab") as fh:
        fh.write(b"x")
    return os.path.getsize(a) + x

File: test-suite/golden-tests/closure-conformance/hosts.rs
use std::rc::Rc;

#[derive(Clone)]
pub struct Holder {
    pub run: std::rc::Rc<dyn rustmorloc::MorlocFn0<i64>>,
    pub tag: i64,
}

#[derive(Clone)]
pub struct Caller {
    pub inc: std::rc::Rc<dyn rustmorloc::MorlocFn1<i64, i64>>,
}

#[derive(Clone)]
pub struct Purer {
    pub apply: std::rc::Rc<dyn rustmorloc::MorlocFn1<i64, i64>>,
}


pub fn home_ident(x: i64) -> i64 {
    x
}

pub fn away_ident(x: i64) -> i64 {
    x
}

pub fn home_tick(path: &String) -> i64 {
    use std::io::Write;
    let mut f = std::fs::OpenOptions::new().append(true).create(true).open(path).unwrap();
    f.write_all(b"x").unwrap();
    std::fs::metadata(path).unwrap().len() as i64
}

pub fn home_mark(s: &String) -> String {
    s.clone()
}

pub fn home_str_len(s: &String) -> i64 {
    s.len() as i64
}

pub fn home_add(x: i64, y: i64) -> i64 {
    x + y
}

pub fn home_fail(path: &String) -> i64 {
    rustmorloc::morloc_throw(format!("failHome ran: {}", path))
}

pub fn home_mk(k: i64) -> std::rc::Rc<dyn rustmorloc::MorlocFn1<i64, i64>> {
    std::rc::Rc::new(move |x: &i64| x + k)
}

pub fn home_big(n: i64) -> Vec<i64> {
    (0..n).collect()
}

pub fn home_use_pure(f: impl rustmorloc::MorlocFn1<i64, i64>, x: i64) -> i64 {
    f.call1(&x)
}

pub fn home_take_thunk(t: impl rustmorloc::MorlocFn0<i64>) -> i64 {
    t.call0();
    t.call0();
    t.call0()
}


pub fn away_bump(path: &String, _x: i64) -> i64 {
    use std::io::Write;
    let mut f = std::fs::OpenOptions::new().append(true).create(true).open(path).unwrap();
    f.write_all(b"x").unwrap();
    std::fs::metadata(path).unwrap().len() as i64
}

pub fn away_take_thunk(t: impl rustmorloc::MorlocFn0<i64>) -> i64 {
    t.call0();
    t.call0();
    t.call0()
}

pub fn away_drop_thunk(_t: impl rustmorloc::MorlocFn0<i64>) -> i64 {
    0
}

pub fn away_use_record(h: &Holder) -> i64 {
    h.run.call0();
    h.run.call0();
    h.run.call0() + h.tag
}

pub fn away_use_list(xs: &Vec<Rc<dyn rustmorloc::MorlocFn0<i64>>>) -> i64 {
    xs[0].call0();
    xs[0].call0();
    xs[0].call0()
}

pub fn away_use_tuple(t: &(Rc<dyn rustmorloc::MorlocFn0<i64>>, i64)) -> i64 {
    t.0.call0();
    t.0.call0();
    t.0.call0() + t.1
}

pub fn away_use_opt(t: &Option<Rc<dyn rustmorloc::MorlocFn0<i64>>>) -> i64 {
    match t {
        Some(f) => {
            f.call0();
            f.call0();
            f.call0()
        }
        None => 0,
    }
}

pub fn away_use_nested(hs: &Vec<Holder>) -> i64 {
    hs[0].run.call0();
    hs[0].run.call0();
    hs[0].run.call0()
}

pub fn away_use_callback(f: impl rustmorloc::MorlocFn1<i64, i64>, x: i64) -> i64 {
    f.call1(&x);
    f.call1(&x);
    f.call1(&x)
}

pub fn away_use_pure(f: impl rustmorloc::MorlocFn1<i64, i64>, x: i64) -> i64 {
    f.call1(&x)
}

pub fn away_use_arity2(f: impl rustmorloc::MorlocFn2<i64, i64, i64>, x: i64, y: i64) -> i64 {
    f.call2(&x, &y)
}

pub fn away_use_pinned(c: &Caller, x: i64) -> i64 {
    c.inc.call1(&x);
    c.inc.call1(&x);
    c.inc.call1(&x)
}

pub fn away_use_hof(
    f: impl rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>, i64>,
    g: impl rustmorloc::MorlocFn1<i64, i64> + 'static,
) -> i64 {
    let boxed: Rc<dyn rustmorloc::MorlocFn1<i64, i64>> = Rc::new(g);
    f.call1(&boxed)
}

pub fn away_use_list_pure(fs: &Vec<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>>, x: i64) -> i64 {
    fs.iter().map(|f| f.call1(&x)).sum()
}

pub fn away_use_mk(
    f: impl rustmorloc::MorlocFn1<i64, Rc<dyn rustmorloc::MorlocFn1<i64, i64>>>,
    a: i64,
    b: i64,
) -> i64 {
    f.call1(&a).call1(&b)
}

pub fn away_use_thunk2(t: impl rustmorloc::MorlocFn0<Rc<dyn rustmorloc::MorlocFn0<i64>>>) -> i64 {
    let inner = t.call0();
    inner.call0();
    inner.call0();
    inner.call0()
}

pub fn away_use_thunk_fn(t: impl rustmorloc::MorlocFn0<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>>, x: i64) -> i64 {
    t.call0().call1(&x)
}

pub fn away_use_fn_thunk(
    f: impl rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn0<i64>>, i64>,
    t: impl rustmorloc::MorlocFn0<i64> + 'static,
) -> i64 {
    let boxed: Rc<dyn rustmorloc::MorlocFn0<i64>> = Rc::new(t);
    f.call1(&boxed)
}

pub fn away_pass_through(f: impl rustmorloc::MorlocFn1<i64, i64> + 'static) -> Rc<dyn rustmorloc::MorlocFn1<i64, i64>> {
    Rc::new(f)
}

pub fn away_map_away<F: Fn(&i64) -> i64>(f: F, xs: &Vec<i64>) -> Vec<i64> {
    xs.iter().map(|x| f(x)).collect()
}

pub fn away_use_hof_eff(
    f: impl rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>, i64>,
    g: impl rustmorloc::MorlocFn1<i64, i64> + 'static,
) -> i64 {
    let boxed: Rc<dyn rustmorloc::MorlocFn1<i64, i64>> = Rc::new(g);
    f.call1(&boxed)
}

pub fn away_use_hof2(
    f: impl rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>, i64>>, i64>,
) -> i64 {
    let h: Rc<dyn rustmorloc::MorlocFn1<Rc<dyn rustmorloc::MorlocFn1<i64, i64>>, i64>> =
        Rc::new(move |g: &Rc<dyn rustmorloc::MorlocFn1<i64, i64>>| g.call1(&3));
    f.call1(&h)
}

pub fn away_two(a: &String, b: &String, _x: i64) -> i64 {
    use std::io::Write;
    let mut f = std::fs::OpenOptions::new().append(true).create(true).open(a).unwrap();
    f.write_all(b"x").unwrap();
    std::fs::metadata(a).unwrap().len() as i64 + b.len() as i64
}

pub fn away_mix(x: i64, a: &String) -> i64 {
    use std::io::Write;
    let mut f = std::fs::OpenOptions::new().append(true).create(true).open(a).unwrap();
    f.write_all(b"x").unwrap();
    std::fs::metadata(a).unwrap().len() as i64 + x
}
